syntax = "proto3";

package perd.workflow.v1;

message AddRequest {
  double a = 1;
  double b = 2;
  string request_id = 3;
  string workflow_name = 4;
}

message AddReply {
  double result = 1;
  string request_id = 2;
  string workflow_name = 3;
  string status_message = 4;
}

message StreamStart {
  string session_id = 1;
  string job_id = 2;
  string workflow_name = 3;
  string operation = 4;
  string params_json = 5;
}

message StreamChunk {
  string session_id = 1;
  string job_id = 2;
  string chunk_id = 3;
  string payload_json = 4;
  bool is_final = 5;
}

message JobCancel {
  string session_id = 1;
  string job_id = 2;
  string reason = 3;
}

message JobStreamRequest {
  oneof payload {
    StreamStart start = 1;
    StreamChunk chunk = 2;
    JobCancel cancel = 3;
  }
}

message JobProgress {
  string session_id = 1;
  string job_id = 2;
  double progress_percent = 3;
  string message = 4;
}

message JobResult {
  string session_id = 1;
  string job_id = 2;
  string result_json = 3;
}

message JobError {
  string session_id = 1;
  string job_id = 2;
  string code = 3;
  string message = 4;
}

message JobStreamResponse {
  oneof event {
    JobProgress progress = 1;
    JobResult result = 2;
    JobError error = 3;
  }
}

service WorkflowService {
  rpc Add(AddRequest) returns (AddReply);
  rpc RunJobStream(stream JobStreamRequest) returns (stream JobStreamResponse);
}
